x-pack/libbeat/management/generate.go

```diff
@@ -78,15 +78,33 @@ func CreateInputsFromStreams(raw *proto.UnitExpectedConfig, inputType string, ag
 		streamSource := raw.GetStreams()[iter].GetSource().AsMap()
 
 		streamSource = injectIndexStream(inputType, raw, stream, streamSource)
-		streamSource, err := injectStreamProcessors(raw, inputType, stream, streamSource)
+
+		// the order of building the processors is important
+		// prepend is used to ensure that the processors defined directly on the stream
+		// come last in the order as they take priority over the others as they are the
+		// most specific to this one stream
+
+		// 1. global processors
+		streamSource = injectGlobalProcesssors(raw, streamSource)
+
+		// 2. agentInfo
+		streamSource, err := injectAgentInfoRule(streamSource, agentInfo)
 		if err != nil {
-			return nil, fmt.Errorf("Error injecting stream processors: %w", err)
+			return nil, fmt.Errorf("Error injecting agent processors: %w", err)
 		}
-		streamSource, err = injectAgentInfoRule(streamSource, agentInfo)
+
+		// 3. stream processors
+		streamSource, err = injectStreamProcessors(raw, inputType, stream, streamSource)
 		if err != nil {
-			return nil, fmt.Errorf("Error injecting agent processors: %w", err)
+			return nil, fmt.Errorf("Error injecting stream processors: %w", err)
 		}
-		streamSource = injectGlobalProcesssors(raw, streamSource)
+
+		// now the order of the processors on this input is as follows
+		// 1. stream processors
+		// 2. agentInfo processors
+		// 3. global processors
+		// 4. stream specific processors
+
```

Comment:
The order here doesn't quite line up with the order of the comments above, which is

1. Global processors
2. agentInfo processors
3. Stream processors

Reply:
It is using prepend not append. So it prepends to the array each time so the order is reversed when reading the code. Operating this way allows the processors defined on the stream to always be last.

It is definitely working as the code comment says and it is working under the Elastic Agent correctly.

Reply:
Yeah was just going to comment that I don't think I read the prepend operations correctly.